[RNMobile] Refactor draggable logic and introduce `DraggableTrigger` component

fluiddot opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Refactors the draggable logic.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The reason for the refactor is mainly to address the following issues spotted during testing:
- Long-press gesture is not working for buttons located within the block's toolbar.
- Long-press gesture is disabled when editing a text on iOS ([reference](https://github.com/WordPress/gutenberg/pull/40101#issuecomment-1091437376)).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

The implementation is described in the following sections:

### Trigger long-press gesture from block component

Originally, the long-press gesture was triggered from a wrapper component over the block list. This was preventing other long-press gestures located in other components to not being triggered, like the inserter button or the up/down arrow buttons within the block's toolbar.

In order to address this issue, a new component named `DraggableTrigger` has been added to handle the long-press gesture. In the following commits you can see the related changes:
- [Refactor draggable component and add `DraggableTrigger` component](https://github.com/WordPress/gutenberg/commit/cb561942b33e00d643525d6094cffe41449d5c51)
- [Use DraggableTrigger in BlockDraggable](https://github.com/WordPress/gutenberg/commit/36c778d8c4d290f2a10d3d322c68961eb6e9617e)

With the introduction of this component, we can now send the `clientId` of the dragged block through the dragging events, so this also implied updating and simplifying the logic of the `BlockDraggable` component and its wrapper.

As an additional note, the `BlockDraggable` component is now rendered in the `BlockListBlock` component, so the dragging gesture doesn't interfere with the block's toolbar ([reference](https://github.com/WordPress/gutenberg/commit/382781d6fdc8a8d9ff726e8d72081ec9d5420046)).

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Drag an unselected block
1. Open a post/page in the app.
2. Add several blocks, including text blocks like the Paragraph block.
3. Long-press on a block that is not selected to activate the dragging mode.
4. Observe that the dragging mode is enabled and that the block can be dragged around the block list.
5. Drop the block and observe that the block gets automatically selected.

### Drag a selected block
1. Open a post/page in the app.
2. Add several blocks, including text blocks like the Paragraph block.
3. Select a block.
4. Long-press on the selected block to activate the dragging mode.
**NOTE:** Only the block's content will be draggable, the block's toolbar won't enable the dragging mode.
5. Observe that the dragging mode is enabled and that the block can be dragged around the block list.
6. Drop the block and observe that the block is still selected.

### Long-press gesture still works when editing a text
1. Open a post/page in the app.
2. Add several blocks, including text blocks like the Paragraph block.
3. Tap on a text block to edit its content.
**NOTE:** Check that the text input is focused (i.e. the text editing is enabled).
4. Long-press on the content.
7. Observe that the long-press gesture works and that doesn't trigger the dragging mode.

### Long-press gesture on toolbar buttons
1. Open a post/page in the app.
2. Long-press on the inserter button (i.e. ➕ button).
3. Observe that a picker is displayed and presents extra options in relation to the target position when inserting the block.
4. Select a block.
5. Long-press on both up and down arrow buttons.
6. Observe that a picker is displayed and presents the expected options.

## Screenshots or screencast <!-- if applicable -->

Android|iOS
--|--
<video src=https://user-images.githubusercontent.com/14905380/163803948-a983a2fe-4680-4d67-82db-8439543b54b5.mp4>|<video src=https://user-images.githubusercontent.com/14905380/163803976-f1659c5a-c580-4a80-8378-c06fe7a6eb6b.MP4>

fluiddot committed:
Present block mover picker only after state updates

fluiddot committed:
Refactor draggable component

fluiddot committed:
Use DraggableTrigger in BlockDraggable

fluiddot committed:
Move BlockDraggable render to BlockListBlock component

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -131,14 +118,15 @@ const BlockDraggableWrapper = ( { children } ) => {
 				srcClientIds: [ currentClientId ],
 				type: 'block',
 			} );
+			selectBlock( currentClientId );
```
The dragged block gets automatically selected after it's dropped.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -261,53 +247,65 @@ const BlockDraggable = ( { clientId, children } ) => {
 			return {
 				isBeingDragged: isBlockBeingDragged( clientId ),
 				isDraggable: 'all' !== templateLock,
+				canDragBlock: hasSelectedBlock()
+					? ! isAnyTextInputFocused
+					: true,
```
We determine if the dragging can be enabled by checking if any text input is focused while a block is selected.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -261,53 +247,65 @@ const BlockDraggable = ( { clientId, children } ) => {
-		} else if ( wasBeingDragged.current ) {
-			stopDraggingBlock();
-			wasBeingDragged.current = false;
+		if ( isBeingDragged !== wasBeingDragged.current ) {
```
The logic of this side effect will only be executed if `isBeingDragged ` value has changed.

fluiddot commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -261,53 +247,65 @@ const BlockDraggable = ( { clientId, children } ) => {
-		<Animated.View style={ wrapperStyles }>
-			{ children( { isDraggable: true } ) }
-		</Animated.View>
+		<DraggableTrigger id={ clientId } enabled={ enabled && canDragBlock }>
```
We now can pass the `clientId` as the id of the draggable component, which will be passed through the dragging events.

fluiddot commented on packages/block-editor/src/components/block-list/block.native.js:
```diff
@@ -254,14 +257,21 @@ class BlockListBlock extends Component {
-							/>
-						) }
+						<BlockDraggable
+							enabled={ ! hasParent }
```
The drag & drop only works for root-level blocks, for this reason, we disable the `BlockDraggable` component when the block has a parent. In the future, once the drag & drop functionality support nested blocks, we could enable this.

fluiddot commented on packages/block-editor/src/components/block-mover/index.native.js:
```diff
@@ -46,9 +47,17 @@ export const BlockMover = ( {
 		}
 
 		setBlockPageMoverState( direction );
-		pickerRef.current.presentPicker();
+		setShouldPresentPicker( true );
```
The options of the picker are controlled by the `blockPageMoverState` state value. We're presenting the picker right after updating the value, however, it's not guaranteed that the picker will present the right options since it might not have been rendered. For this reason, I updated this logic and now the presentation of the picker is controlled with another state value.

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -1,103 +1,171 @@
-		.onEnd( () => {
-			'worklet';
-			if ( isAnyTextInputFocused.value ) {
+	useAnimatedReaction(
+		() => isDragging.value,
```
The dragging evens are triggered by the `isDragging` shared value, which is exposed via the provider.

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -1,103 +1,171 @@
 	const panGesture = Gesture.Pan()
 		.manualActivation( true )
+		.onTouchesDown( ( event ) => {
+			const { x = 0, y = 0 } = event.allTouches[ 0 ];
```
We get the touch data from the first finger put over the screen.

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -1,103 +1,171 @@
+			initialPosition.y.value = y;
+		} )
 		.onTouchesMove( ( _, state ) => {
-			'worklet';
```
By default the handler functions are treated as worklets, so looks like it's not necessary to marked as that.

fluiddot committed:
Fix long-press gesture when editing a text on iOS

fluiddot committed:
Memoize draggable provider value to prevent re-renders

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -1,103 +1,171 @@
+			enabled={ enabled }
+			minDurationMs={ minDuration }
+			maxDist={ maxDistance }
+			simultaneousHandlers={ panGestureRef }
```
This props is required to allow the pan gesture to be triggered along with the long-press.

fluiddot commented on packages/components/src/draggable/index.native.js:
```diff
@@ -1,103 +1,173 @@
 		.shouldCancelWhenOutside( false );
 
-	const dragHandler = Gesture.Simultaneous( panGesture, longPressGesture );
+	const providerValue = useMemo( () => {
+		return { panGestureRef, isDragging, draggingId };
+	}, [] );
```
The provider component would cause a re-render if we pass this object in the `value` prop, as it would create a new instance on every render. In order to prevent this, the value is memoized. Note that the dependency is array is empty as all values referenced don't change after the component is mounted:
- `panGestureRef` is a React reference, only the value of `current` will change which is accessed by reference.
- `isDragging` and `draggingId` are shared values whose values are accessed by reference via the `value` property.

geriux commented:
Hey there 👋 great work with the refactor! I'm still testing and reviewing the code but so far encountered a bug on iOS while scrolling:

https://user-images.githubusercontent.com/4885740/163971542-dac70049-71e8-4824-9088-a465df142269.mov

It looks like there's an issue with the scroll, in the video, I'm scrolling really fast but I've tried scrolling a little slower and it's the same, the dragging functionality stays activated showing the chip and the indicator. Trying to scroll doesn't work and the only way to make it work again is to try to activate the dragging in a block 😅 

This only happens on iOS, on Android is working fine.

Edit: By the way this happens on both simulator and device

fluiddot commented:
> Hey there 👋 great work with the refactor! I'm still testing and reviewing the code but so far encountered a bug on iOS while scrolling:
> 
>  ScrollingBug.mov 
> It looks like there's an issue with the scroll, in the video, I'm scrolling really fast but I've tried scrolling a little slower and it's the same, the dragging functionality stays activated showing the chip and the indicator. Trying to scroll doesn't work and the only way to make it work again is to try to activate the dragging in a block 😅
> 
> This only happens on iOS, on Android is working fine.

@geriux Thanks for sharing the issue 🙇! I'll take a deeper look in case I overlooked anything during the refactor 👀 .

fluiddot commented:
> > Hey there 👋 great work with the refactor! I'm still testing and reviewing the code but so far encountered a bug on iOS while scrolling:
> > ScrollingBug.mov
> > It looks like there's an issue with the scroll, in the video, I'm scrolling really fast but I've tried scrolling a little slower and it's the same, the dragging functionality stays activated showing the chip and the indicator. Trying to scroll doesn't work and the only way to make it work again is to try to activate the dragging in a block 😅
> > This only happens on iOS, on Android is working fine.
> 
> @geriux Thanks for sharing the issue 🙇! I'll take a deeper look in case I overlooked anything during the refactor 👀 .

As far as I checked, on iOS, the [property `maxDist` (i.e. maximum distance)](https://docs.swmansion.com/react-native-gesture-handler/docs/2.3.0/gesture-handlers/api/longpress-gh#maxdist) of the long-press gesture can make it fail even when it's active if you move the finger further than the distance set. Per the documentation of the property, this shouldn't occur as the maximum distance should only produce a failure if the gesture is not active 🙃 .

> Maximum distance, expressed in points, that defines how far the finger is allowed to travel during a long press gesture. If the finger travels further than the defined distance and the handler hasn't yet [activated](https://docs.swmansion.com/react-native-gesture-handler/docs/2.3.0/gesture-handlers/basics/state#active), it will fail to recognize the gesture.

However, this is not the case for iOS which I confirm that it produces the gesture to fail.

As a workaround, I'm going to listen for the `end` event of the pan gesture and disable dragging in that case.

fluiddot committed:
Fix dragging not being disabled after scrolling

fluiddot commented:
@geriux Regarding the issue you spotted and outlined in [this comment](https://github.com/WordPress/gutenberg/pull/40406#issuecomment-1102368363), I've just pushed a fix in https://github.com/WordPress/gutenberg/pull/40406/commits/06fd88a965cf92bfbac7dd897bc72948e6889d19 to address it. I tested locally on both platforms and looks like it can no longer be reproduced, in any case, let me know if you consider it fixed and of course, if you find anything while testing. Thanks 🙇 !

fluiddot committed:
Reduce calls to event handlers of pan and long-press gestures

fluiddot committed:
Prevent onDragEnd event to be called upon mounting

fluiddot commented:
@geriux I've just pushed a couple of commits to fix the following issues I spotted while testing:

#### [Prevent onDragEnd event to be called upon mounting](https://github.com/WordPress/gutenberg/pull/40406/commits/9fe4fcfee2260c1ac6ebe0501aeefc38955fb85a)
I noticed when adding console logs that the `onDragEnd` event was always being called when opening the editor. This doesn't produce any issue as no block is dragged, but I fixed to prevent potential side effects in the future.

#### [Reduce calls to event handlers of pan and long-press gestures](https://github.com/WordPress/gutenberg/pull/40406/commits/4f861dc509eca217bca2643337f79cbecab6b122)
Similar to the previous issue, I spotted that some of the event handlers of tap and long-press gestures were being called several times (actually on every frame) which is not expected, specifically the `onTouchesMove` event of pan gesture and `onActive` of long-press gesture. This behavior doesn't impact the dragging logic but I fixed them to improve performance.

`OnTouchesMove` event of pan gesture is expected that's called several times as it's triggered every time the finger moves, but since we only use it for enabling the gesture after the long-press gesture, I introduced a fix to ensure the logic is only executed once.

For long-press gesture's `onActive`, I initially thought that `onActive` would be triggered only once (when transitioning to the `active` state), however, I saw that it's called on every frame 🙃 . Similarly to the pan gesture, I added a fix to ensure that it's only executed once.

geriux commented:
> @geriux I've just pushed a couple of commits to fix the following issues I spotted while testing:
> 
> #### [Prevent onDragEnd event to be called upon mounting](https://github.com/WordPress/gutenberg/pull/40406/commits/9fe4fcfee2260c1ac6ebe0501aeefc38955fb85a)
> I noticed when adding console logs that the `onDragEnd` event was always being called when opening the editor. This doesn't produce any issue as no block is dragged, but I fixed to prevent potential side effects in the future.
> 
> #### [Reduce calls to event handlers of pan and long-press gestures](https://github.com/WordPress/gutenberg/pull/40406/commits/4f861dc509eca217bca2643337f79cbecab6b122)
> Similar to the previous issue, I spotted that some of the event handlers of tap and long-press gestures were being called several times (actually on every frame) which is not expected, specifically the `onTouchesMove` event of pan gesture and `onActive` of long-press gesture. This behavior doesn't impact the dragging logic but I fixed them to improve performance.
> 
> `OnTouchesMove` event of pan gesture is expected that's called several times as it's triggered every time the finger moves, but since we only use it for enabling the gesture after the long-press gesture, I introduced a fix to ensure the logic is only executed once.
> 
> For long-press gesture's `onActive`, I initially thought that `onActive` would be triggered only once (when transitioning to the `active` state), however, I saw that it's called on every frame 🙃 . Similarly to the pan gesture, I added a fix to ensure that it's only executed once.

Awesome! Its good to have those checks. Thanks for the changes! I'll continue with the review and testing 🚀

geriux commented on packages/block-editor/src/components/block-draggable/index.native.js:
```diff
@@ -261,53 +248,76 @@ const BlockDraggable = ( { clientId, children } ) => {
+		<DraggableTrigger
+			id={ clientId }
+			enabled={ enabled && canDragBlock }
+			minDuration={ Platform.select( {
```
What do you think if we move this to a variable? 😁

geriux replied:
Also can we have a const for the `450` value?

fluiddot replied:
For the `450` value, I'm considering calculating it by subtracting `50` ms from `DEFAULT_LONG_PRESS_MIN_DURATION`, as we only need to have a lower value.

geriux replied:
That sounds good, I think having that calculation in a variable would be good, like:

```javascript
const DEFAULT_IOS_LONG_PRESS_MIN_DURATION = DEFAULT_LONG_PRESS_MIN_DURATION - 50;
```

Or something like that 😄

fluiddot replied:
I applied this suggestion in https://github.com/WordPress/gutenberg/pull/40406/commits/61f6c7224c1db7ab4c29af97d7bed345a8b60427.

geriux reviewed (commented):
Hey @fluiddot 👋 

After some testing I think I found another bug:

https://user-images.githubusercontent.com/4885740/164042981-e7deaf9e-a783-4350-9b7b-91d8df72fd7d.mov

The steps I followed:

- Move a text-based block (Paragraph, heading)
- Edit the block's text content
- Press the enter key to add another Paragraph block
- Select another text-based block and try to drag & drop it (outside the text input)

It's happening on both iOS and Android.

fluiddot commented:
@geriux looks like I had some comments within the code pending publication 😅, sorry for posting them now. Hopefully, they will help us with future reviews.

fluiddot committed:
Add DEFAULT_IOS_LONG_PRESS_MIN_DURATION constant

fluiddot commented:
> Hey @fluiddot 👋
> 
> After some testing I think I found another bug:
> 
>  TextEditing.mov 
> The steps I followed:
> 
> * Move a text-based block (Paragraph, heading)
> * Edit the block's text content
> * Press the enter key to add another Paragraph block
> * Select another text-based block and try to drag & drop it (outside the text input)
> 
> It's happening on both iOS and Android.

@geriux I'm thinking of addressing this bug in a different PR where I'd like also to tackle https://github.com/wordpress-mobile/gutenberg-mobile/issues/4775. This way we allow the next tasks to use these changes, wdyt?

geriux commented:
> @geriux I'm thinking of addressing this bug in a different PR where I'd like also to tackle [wordpress-mobile/gutenberg-mobile#4775](https://github.com/wordpress-mobile/gutenberg-mobile/issues/4775). This way we allow the next tasks to use these changes, wdyt?

Sounds good! Thanks!

geriux reviewed (approved):
LGTM! Really nice work with this refactor! 👏  Tested it on both iOS and Android and it's working well except for the reported bug that will be handled [as agreed](https://github.com/WordPress/gutenberg/pull/40406#issuecomment-1103630441) in another PR.

fluiddot commented:
Looks like we need to include changes from https://github.com/WordPress/gutenberg/commit/1e06cba9287a964505bb84b05500c558d919ed71 to make `Unit Tests / PHP` PR check pass. I'll update the upstream branch [rnmobile/feature/drag-and-drop](https://github.com/WordPress/gutenberg/tree/rnmobile/feature/drag-and-drop) with the latest changes from `trunk`.
